libmachine/provision/redhat.go

```diff
@@ -0,0 +1,279 @@
+	})
+}
+
+func NewRedHatProvisioner(d drivers.Driver) Provisioner {
+	return &RedHatProvisioner{
+		packages: []string{
+			"curl",
+		},
+		Driver: d,
+	}
+}
+
+type RedHatProvisioner struct {
+	packages      []string
+	OsReleaseInfo *OsRelease
+	Driver        drivers.Driver
+	AuthOptions   auth.AuthOptions
+	EngineOptions engine.EngineOptions
+	SwarmOptions  swarm.SwarmOptions
+}
+
+func (provisioner *RedHatProvisioner) Service(name string, action pkgaction.ServiceAction) error {
+	command := fmt.Sprintf("sudo systemctl %s %s", action.String(), name)
+
+	if _, err := provisioner.SSHCommand(command); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *RedHatProvisioner) Package(name string, action pkgaction.PackageAction) error {
+	var packageAction string
+
+	switch action {
+	case pkgaction.Install:
+		packageAction = "install"
+	case pkgaction.Remove:
+		packageAction = "remove"
+	case pkgaction.Upgrade:
+		packageAction = "upgrade"
+	}
+
+	switch name {
+	case "docker":
+		name = "docker"
+	}
+
+	command := fmt.Sprintf("sudo -E yum %s -y %s", packageAction, name)
+
+	if _, err := provisioner.SSHCommand(command); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *RedHatProvisioner) isAWS() bool {
+	if _, err := provisioner.SSHCommand("curl -s http://169.254.169.254/latest/meta-data/ami-id"); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (provisioner *RedHatProvisioner) configureRepos() error {
+
+	// TODO: should this be handled differently? on aws we need to enable
+	// the extras repo different than a standalone rhel box
+
+	repoCmd := "subscription-manager repos --enable=rhel-7-server-extras-rpms"
+	if provisioner.isAWS() {
+		repoCmd = "yum-config-manager --enable rhui-REGION-rhel-server-extras"
+	}
+
+	if _, err := provisioner.SSHCommand(fmt.Sprintf("sudo %s", repoCmd)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *RedHatProvisioner) installDocker() error {
+	if err := provisioner.Package("docker", pkgaction.Install); err != nil {
+		return err
+	}
+
+	if err := provisioner.Service("docker", pkgaction.Restart); err != nil {
+		return err
+	}
+
+	if err := provisioner.Service("docker", pkgaction.Enable); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *RedHatProvisioner) dockerDaemonResponding() bool {
+	if _, err := provisioner.SSHCommand("sudo docker version"); err != nil {
+		log.Warn("Error getting SSH command to check if the daemon is up: %s", err)
+		return false
+	}
+
+	// The daemon is up if the command worked.  Carry on.
+	return true
+}
+
+func (provisioner *RedHatProvisioner) Provision(swarmOptions swarm.SwarmOptions, authOptions auth.AuthOptions, engineOptions engine.EngineOptions) error {
+	provisioner.SwarmOptions = swarmOptions
+	provisioner.AuthOptions = authOptions
+	provisioner.EngineOptions = engineOptions
+
+	// set default storage driver for redhat
+	provisioner.EngineOptions.StorageDriver = "devicemapper"
```

Comment:
Actually, we might want to default to only setting this if the option is empty: If the user knows what they are doing and wants to over-ride the default, we shouldn't prevent them (this would).